src/Common/src/TypeSystem/IL/Stubs/TypeGetTypeMethodThunk.cs

```diff
@@ -0,0 +1,163 @@
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return OwningType.Context;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return $"{_helperMethod.Name}_{Signature.Length}_{DefaultAssemblyName}";
+            }
+        }
+
+        public override TypeDesc OwningType
+        {
+            get;
+        }
+
+        public override MethodSignature Signature
+        {
+            get;
+        }
+
+        public string DefaultAssemblyName
+        {
+            get;
+        }
+
+        public override MethodIL EmitIL()
+        {
+            ILEmitter emit = new ILEmitter();
+            ILCodeStream codeStream = emit.NewCodeStream();
+
+            Debug.Assert(Signature[0] == _helperMethod.Signature[0]);
+            codeStream.EmitLdArg(0);
+
+            Debug.Assert(_helperMethod.Signature[1].IsString);
+            codeStream.Emit(ILOpcode.ldstr, emit.NewToken(DefaultAssemblyName));
+
+            for (int i = 2; i < _helperMethod.Signature.Length; i++)
+            {
+                // The helper method could be expecting more arguments than what we have - check for that
+                // The thunk represents one of the 6 possible overloads:
+                // (String), (String, bool), (String, bool, bool)
+                // (String, Func<...>, Func<...>), (String, Func<...>, Func<...>, bool), (String, Func<...>, Func<...>, bool, bool)
+                // We only need 2 helpers to support all 6 overloads. The default value for the bools is false.
+
+                if (i - 1 < Signature.Length)
+                {
+                    // Pass user's parameter
+                    Debug.Assert(_helperMethod.Signature[i] == Signature[i - 1]);
+                    codeStream.EmitLdArg(i - 1);
+                }
+                else
+                {
+                    // Pass a default value
+                    Debug.Assert(_helperMethod.Signature[i].IsWellKnownType(WellKnownType.Boolean));
+                    codeStream.EmitLdc(0);
+                }
+            }
+
+            codeStream.Emit(ILOpcode.call, emit.NewToken(_helperMethod));
+            codeStream.Emit(ILOpcode.ret);
+
+            return emit.Link(this);
+        }
+    }
+
+    internal abstract class TypeGetTypeMethodThunkCache
+    {
+        private TypeDesc _owningTypeForThunks;
+        private Unifier _cache;
+
+        public TypeGetTypeMethodThunkCache(TypeDesc owningTypeForThunks)
+        {
+            _owningTypeForThunks = owningTypeForThunks;
+            _cache = new Unifier(this);
+        }
+
+        public MethodDesc GetHelper(MethodDesc getTypeOverload, string defaultAssemblyName)
+        {
+            return _cache.GetOrCreateValue(new Key(defaultAssemblyName, getTypeOverload));
+        }
+
+        protected abstract MethodDesc GetHelperForOverload(MethodDesc typeGetTypeOverload);
```

Comment:
Mmm, would making a forwarder for these helpers in `Internal.Runtime.CompilerHelpers` in CoreLib (that uses the callback mechanism to get to Reflection.Execution) still be overengineered?

What I'm fighting with is the hardcoding of the composition of the base class library in the compiler.